src/Interpreters/ProcessList.cpp

```diff
@@ -297,6 +297,13 @@ ProcessListEntry::~ProcessListEntry()
         }
     }
 
+    /// Wait for the query if it is in the cancellation right now.
+    parent.cancelled.cv.wait(lock.lock, [&]()
+    {
+        auto & processes = parent.cancelled.processes;
+        return std::find(processes.begin(), processes.end(), &*it) == processes.end();
+    });
```

Comment:
Could that method be the member function of CancelledProcesses?

Reply:
This looks like a local thing, I don't think that it worth to move it into the CancelledProcesses.
And actually that CancelledProcesses was added only to group related fields together.